src/plugin.ts

```diff
@@ -433,28 +433,34 @@ export class MongoDBOIDCPluginImpl implements MongoDBOIDCPlugin {
     // for client A, the token expires before it is requested again by client A,
     // then the plugin is passed to client B which requests a token, and we
     // receive mismatching tokens for different users or different audiences.
-    const idTokenClaims = tokenSet.claims();
-    if (state.lastIdTokenClaims) {
-      for (const claim of ['aud', 'sub'] as const) {
-        const normalize = (value: string | string[]): string => {
-          return JSON.stringify(
-            Array.isArray(value) ? [...value].sort() : value
-          );
-        };
-        const knownClaim = normalize(state.lastIdTokenClaims[claim]);
-        const newClaim = normalize(idTokenClaims[claim]);
-
-        if (knownClaim !== newClaim) {
-          throw new MongoDBOIDCError(
-            `Unexpected '${claim}' field in id token: Expected ${knownClaim}, saw ${newClaim}`
-          );
+    if (tokenSet.id_token) {
+      const idTokenClaims = tokenSet.claims();
+      if (state.lastIdTokenClaims) {
```

Comment:
There are two possible reasons why this branch would not be taken – either there was no previous token set that we could have received (i.e. this is the first one), or there have already received a previous token set that did not contain an ID token. In the latter case, we should throw just like we do in the `else if` branch below, because then the tokens mismatch. 

So I'd say we need to store another bit of information here – either give `state.lastIdTokenClaims` a special value that indicates "no ID token received", or add a property to the value of `state.lastIdTokenClaims`, or maybe even just translate "no ID token received" to "ID token with empty/null `aud` and `sub` received" (which could never happen if we did receive an ID token). No strong feelings about which of these make the most sense, though.

Reply:
thanks, I totally missed that! looks like unit tests would be in order actually 😅  I moved the errors up to make it a bit easier to read (hopefully). it felt weird to extend the `lastIdTokenClaims?: Pick<IdTokenClaims, 'aud' | 'sub'>;` type to cover the no id token case, in the end I put it as a separate tag. the new tag is set together with the lastIdToken claims and added also to serialize and deserialize functions.